File: api/scripts/threedivi-liveness.py
#!/usr/bin/env python3
"""
3DiVi Face SDK liveness (Python Processing Block API).
Usage: python3 threedivi-liveness.py <sdk_path> <modification> <image1.jpg> [image2.jpg ...]
"""
from __future__ import annotations

import json
import os
import sys

if len(sys.argv) < 4:
    print("Usage: threedivi-liveness.py <sdk_path> <modification> <image>...", file=sys.stderr)
    sys.exit(2)

sdk_path = sys.argv[1]
modification = sys.argv[2]
image_paths = sys.argv[3:]
min_confidence = float(os.environ.get("THREEDIVI_LIVENESS_MIN_CONFIDENCE", "0.5"))

sys.path.insert(0, os.path.join(sdk_path, "python_api"))

from face_sdk_3divi import FacerecService  # noqa: E402

dll_path = os.path.join(sdk_path, "lib", "libfacerec.so")
conf_path = os.path.join(sdk_path, "conf", "facerec")
license_path = os.path.join(sdk_path, "license")

service = FacerecService.create_service(dll_path, conf_path, license_path)

detector = service.create_processing_block(
    {"unit_type": "FACE_DETECTOR", "modification": "ssyv_light", "use_cuda": False}
)
fitter = service.create_processing_block(
    {"unit_type": "FACE_FITTER", "modification": "fda", "use_cuda": False}
)
liveness_estimator = service.create_processing_block(
    {
        "unit_type": "LIVENESS_ESTIMATOR",
        "modification": modification,
        "use_cuda": False,
    }
)


def read_image(path: str) -> bytes:
    with open(path, "rb") as f:
        return f.read()


best = {"passed": False, "verdict": "UNKNOWN", "confidence": 0.0, "faces_detected": 0}

for image_path in image_paths:
    data = service.create_context_from_encoded_image(read_image(image_path))
    detector(data)

    objects = list(data["objects"])
    if len(objects) != 1:
        continue

    fitter(data)
    liveness_estimator(data)

    for obj in objects:
        verdict = obj["liveness"]["value"].get_string()
        confidence = float(obj["liveness"]["confidence"].get_double())
        passed = verdict == "REAL" and confidence >= min_confidence

        if passed and confidence > best["confidence"]:
            best = {"passed": True, "verdict": verdict, "confidence": confidence, "faces_detected": 1}
        elif not best["passed"] and confidence > best["confidence"]:
            best = {
                "passed": passed,
                "verdict": verdict,
                "confidence": confidence,
                "faces_detected": 1,
            }

print(json.dumps(best))

File: api/scripts/threedivi-liveness.mjs
#!/usr/bin/env node
/**
 * 3DiVi Face SDK liveness check (Node.js API).
 * Usage: node threedivi-liveness.mjs <sdk_path> <modification> <image1.jpg> [image2.jpg ...]
 * Requires THREEDIVI_NODE_API_PATH or <sdk_path>/../node_js_api from face-sdk release.
 */
import fs from 'fs';
import path from 'path';
import { createRequire } from 'module';

const [sdkPath, modification, ...imagePaths] = process.argv.slice(2);

if (!sdkPath || !modification || imagePaths.length === 0) {
  console.error('Usage: threedivi-liveness.mjs <sdk_path> <modification> <image>...');
  process.exit(2);
}

const nodeApiPath =
  process.env.THREEDIVI_NODE_API_PATH?.trim() ||
  path.join(sdkPath, 'node_js_api');

const require = createRequire(import.meta.url);
const facerec = require(path.join(nodeApiPath, 'face_sdk_3divi.js'));

const dllPath = path.join(sdkPath, 'lib', process.platform === 'darwin' ? 'libfacerec.dylib' : 'libfacerec.so');
const confPath = path.join(sdkPath, 'conf', 'facerec');
const licensePath = path.join(sdkPath, 'license');

facerec.Init(dllPath, confPath, licensePath);

const detectorConfig = new facerec.Context();
detectorConfig.get('unit_type').value = 'FACE_DETECTOR';
detectorConfig.get('modification').value = 'ssyv_light';

const fitterConfig = new facerec.Context();
fitterConfig.get('unit_type').value = 'FACE_FITTER';

const livenessConfig = new facerec.Context();
livenessConfig.get('unit_type').value = 'LIVENESS_ESTIMATOR';
livenessConfig.get('modification').value = modification;

const detector = new facerec.ProcessingBlock(detectorConfig);
const fitter = new facerec.ProcessingBlock(fitterConfig);
const livenessEstimator = new facerec.ProcessingBlock(livenessConfig);

const minConfidence = Number(process.env.THREEDIVI_LIVENESS_MIN_CONFIDENCE ?? '0.5');

let best = { passed: false, verdict: 'UNKNOWN', confidence: 0, faces_detected: 0 };

for (const imagePath of imagePaths) {
  const buffer = fs.readFileSync(imagePath);
  const ioData = new facerec.Context(buffer);

  detector.process(ioData);
  const objects = ioData.get('objects');
  const faceCount = objects?.length ?? 0;

  if (faceCount !== 1) {
    continue;
  }

  fitter.process(ioData);
  livenessEstimator.process(ioData);

  const obj = objects[0];
  const liveness = obj.get('liveness');
  const verdict = liveness.get('value').value;
  const confidence = Number(liveness.get('confidence').value);
  const passed = verdict === 'REAL' && confidence >= minConfidence;

  if (passed && confidence > best.confidence) {
    best = { passed: true, verdict, confidence, faces_detected: 1 };
  } else if (!best.passed && confidence > best.confidence) {
    best = { passed, verdict, confidence, faces_detected: 1 };
  }
}

console.log(JSON.stringify(best));
process.exit(0);
